daemon/src/common/Assert.h

```diff
@@ -0,0 +1,159 @@
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+===========================================================================
+*/
+
+#ifndef COMMON_ASSERT_H_
+#define COMMON_ASSERT_H_
+
+/*
+ * Daemon asserts to be used instead of the regular C stdlib assert function (if you don't
+ * use assert yet, you should start now!). In debug ASSERT(condition) will trigger an error,
+ * otherwise in release it does nothing at runtime.
+ *
+ * If possible use the ASSERT_(EQ|NQ|LT|LE|GT|GE)(a, b) variants that are respectively
+ * equivalent to ASSERT(a OP b) with being ==, !=, <, <=, >, >=. These variants are better
+ * because they print the values of the operands in the error message, so you have more
+ * information to start debugging when the assert fires.
+ *
+ * In case of name clashes (with for example a testing library), you can define the
+ * DEAMON_SKIP_ASSERT_SHORTHANDS to only define the DAEMON_ prefixed macros.
+ *
+ * These asserts feature:
+ *     - Logging of the error with file, line and function information.
+ *     - The printing of operands when using the specialized versions.
+ *     - Breaking in the debugger when an assert is triggered and a debugger is attached.
+ *     - Use the assert information to help the compiler optimizer in release builds.
+ */
+
+// MSVC triggers a warning in /W4 for do {} while(0). SDL worked around this by using
+// (0,0) and points out that it looks like an owl face.
+#if defined(_MSC_VER)
+    #define DAEMON_ASSERT_LOOP_CONDITION (0,0)
+#else
+    #define DAEMON_ASSERT_LOOP_CONDITION (0)
+#endif
+
+// DAEMON_ASSERT_CALLSITE_HELPER generates the actual assert code. In Debug it does what you would
+// expect of an assert and in release it tries to give hints to make the compiler generate better code.
+#if defined(DEBUG)
+    #define DAEMON_ASSERT_CALLSITE_HELPER(file, func, line, _0, code, condition, message) \
+        do { \
+            code; \
+            if (!(condition)) { \
+                std::string msg = Str::Format("Assertion failure at %s:%s (%s): %s", file, line, func, message); \
+                if (Sys::IsDebuggerAttached()) { \
+                    Log::Warn(msg); \
+                    BREAKPOINT(); \
+                } \
+                Sys::Error(msg); \
+            } \
+        } while(DAEMON_ASSERT_LOOP_CONDITION)
+#else
+    #if defined(_MSC_VER)
+        #define DAEMON_ASSERT_CALLSITE_HELPER(file, func, line, directExpression, _1, _2, _3) \
+            __assume(directExpression)
+    #elif defined(__GNUC__)
+        #define DAEMON_ASSERT_CALLSITE_HELPER(file, func, line, directExpression, _1, _2, _3) \
+            if (!(directExpression)) { \
+                __builtin_unreachable(); \
+            }
```

Comment:
Yes, it will generate code equivalent to compiler hints for optimization. Looking at it now I realize on GCC and clang it might call function calls. I'll need to find a better builtin if one exists.